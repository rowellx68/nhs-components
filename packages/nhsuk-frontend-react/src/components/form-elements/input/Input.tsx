import React, { ReactNode } from 'react';
import {
  BaseFormElementProps,
  FormGroup,
} from '@/components/core/form-group/FormGroup';
import { Factory, factory } from '@/internal/factory/factory';
import { ColumnWidth, ElementProps, InputWidth } from '@/types/shared';
import clsx from 'clsx';
import { Base } from '@/components/core/base/Base';

export type InputProps = {
  width?: InputWidth | ColumnWidth;
  prefix?: ReactNode;
  suffix?: ReactNode;
} & BaseFormElementProps &
  ElementProps<'input', 'size' | 'as'>;

type InputFactory = Factory<{
  props: InputProps;
  ref: HTMLInputElement;
}>;

const Input = factory<InputFactory>(
  ({ width, suffix, prefix, ...props }, ref) => {
    const component = suffix || prefix ? 'div' : React.Fragment;
    const baseProps =
      component === 'div' ? { className: 'nhsuk-input__wrapper' } : {};

    const widthNumberLike = width && !isNaN(parseInt(width, 10));

    return (
      <FormGroup
        as="input"
        withErrorLine
        {...props}
        ref={ref}
        inputType="input"
        render={({ id, name, className, withError, ...rest }) => (
          <Base<any> as={component} {...baseProps}>
            {prefix && (
              <div className="nhsuk-input__prefix" aria-hidden="true">
                {prefix}
              </div>
            )}
            <input
              id={id}
              name={name}
              className={clsx(
                'nhsuk-input',
                {
                  [`nhsuk-input--width-${width}`]: widthNumberLike,
                  [`nhsuk-u-width-${width}`]: !widthNumberLike && width,
                  'nhsuk-u-width-full': !widthNumberLike && !width,
                  'nhsuk-input--error': withError,
                },
                className,
              )}
              {...rest}
            />
            {suffix && (
              <div className="nhsuk-input__suffix" aria-hidden="true">
                {suffix}
              </div>
            )}
          </Base>
        )}
      />
    );
  },
);

Input.displayName = 'Input';

export { Input };
