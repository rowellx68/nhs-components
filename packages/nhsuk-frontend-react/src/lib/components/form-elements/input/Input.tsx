import FormGroup from '@/components/core/form-group/FormGroup'
import { FormElementProps } from '@/types/form-element-types'
import { InputWidth } from '@/types/nhsuk-sizes'
import clsx from 'clsx'
import {
  ForwardRefRenderFunction,
  HTMLInputTypeAttribute,
  HTMLProps,
  ReactNode,
  forwardRef,
} from 'react'

type InputProps = {
  width?: InputWidth
  type?: HTMLInputTypeAttribute
  prefix?: ReactNode
  suffix?: ReactNode
} & Omit<HTMLProps<HTMLInputElement>, 'type'> &
  FormElementProps

/**
 * Use text input to let users enter a single line of text.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/text-input).
 *
 * @example
 * ```tsx
 * <Input id="example" label="Example" />
 * ```
 */
const Input: ForwardRefRenderFunction<HTMLInputElement, InputProps> = (
  props,
  ref,
): JSX.Element => {
  return (
    <FormGroup<InputProps>
      {...props}
      inputType="input"
      render={({ id, name, className, error, width, type = 'text', ...rest }) =>
        props.prefix || props.suffix ? (
          <div className="nhsuk-input__wrapper">
            {props.prefix && (
              <div className="nhsuk-input__prefix" aria-hidden="true">{props.prefix}</div>
            )}
            <input
              id={id}
              name={name}
              className={clsx(
                'nhsuk-input',
                {
                  [`nhsuk-input--width-${width}`]: width,
                  'nhsuk-input--error': error,
                },
                className,
              )}
              type={type}
              ref={ref}
              {...rest}
            />
            {props.suffix && (
              <div className="nhsuk-input__suffix" aria-hidden="true">{props.suffix}</div>
            )}
          </div>
        ) : (
          <input
            id={id}
            name={name}
            className={clsx(
              'nhsuk-input',
              {
                [`nhsuk-input--width-${width}`]: width,
                'nhsuk-input--error': error,
              },
              className,
            )}
            type={type}
            ref={ref}
            {...rest}
          />
        )
      }
    />
  )
}

const InputForwardRef = forwardRef(Input)

InputForwardRef.displayName = 'Input'

export default InputForwardRef
