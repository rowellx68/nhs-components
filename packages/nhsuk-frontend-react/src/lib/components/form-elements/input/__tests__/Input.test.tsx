import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import Input from '..'

describe('Input', () => {
  test('should match the snapshot', () => {
    render(<Input data-testid="input" />)

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - label', () => {
    render(<Input data-testid="input" label="First name" />)

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - hint', () => {
    render(
      <Input data-testid="input" hint="Optionally provide your phone number" />,
    )

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - error', () => {
    render(<Input data-testid="input" error="This is an error" />)

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - type password', () => {
    render(<Input data-testid="input" type="password" />)

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - prefix', () => {
    render(<Input data-testid="input" prefix="Â£" />)

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - suffix', () => {
    render(<Input data-testid="input" suffix="mmHg" />)

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })
})
