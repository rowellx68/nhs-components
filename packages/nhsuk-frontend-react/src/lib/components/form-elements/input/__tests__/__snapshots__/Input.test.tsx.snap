// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > should match the snapshot - error 1`] = `
<div
  class="nhsuk-form-group nhsuk-form-group--error"
>
  <span
    class="nhsuk-error-message"
    for="input_:r3:"
    id="input_:r3:--error-message"
  >
    <span
      class="nhsuk-u-visually-hidden"
    >
      Error
      :
    </span>
    This is an error
  </span>
  <input
    class="nhsuk-input nhsuk-input--error"
    data-testid="input"
    id="input_:r3:"
    name="input_:r3:"
    type="text"
  />
</div>
`;

exports[`Input > should match the snapshot - hint 1`] = `
<div
  class="nhsuk-form-group"
>
  <div
    class="nhsuk-hint"
    for="input_:r2:"
    id="input_:r2:--hint"
  >
    Optionally provide your phone number
  </div>
  <input
    aria-describedby="input_:r2:--hint"
    class="nhsuk-input"
    data-testid="input"
    id="input_:r2:"
    name="input_:r2:"
    type="text"
  />
</div>
`;

exports[`Input > should match the snapshot - label 1`] = `
<div
  class="nhsuk-form-group"
>
  <label
    class="nhsuk-label"
    for="input_:r1:"
    id="input_:r1:--label"
  >
    First name
  </label>
  <input
    aria-labelledby="input_:r1:--label"
    class="nhsuk-input"
    data-testid="input"
    id="input_:r1:"
    name="input_:r1:"
    type="text"
  />
</div>
`;

exports[`Input > should match the snapshot - prefix 1`] = `
<div
  class="nhsuk-input__wrapper"
>
  <div
    aria-hidden="true"
    class="nhsuk-input__prefix"
  >
    £
  </div>
  <input
    class="nhsuk-input"
    data-testid="input"
    id="input_:r5:"
    name="input_:r5:"
    prefix="£"
    type="text"
  />
</div>
`;

exports[`Input > should match the snapshot - suffix 1`] = `
<div
  class="nhsuk-input__wrapper"
>
  <input
    class="nhsuk-input"
    data-testid="input"
    id="input_:r6:"
    name="input_:r6:"
    suffix="mmHg"
    type="text"
  />
  <div
    aria-hidden="true"
    class="nhsuk-input__suffix"
  >
    mmHg
  </div>
</div>
`;

exports[`Input > should match the snapshot - type password 1`] = `
<div
  class="nhsuk-form-group"
>
  <input
    class="nhsuk-input"
    data-testid="input"
    id="input_:r4:"
    name="input_:r4:"
    type="password"
  />
</div>
`;

exports[`Input > should match the snapshot 1`] = `
<div
  class="nhsuk-form-group"
>
  <input
    class="nhsuk-input"
    data-testid="input"
    id="input_:r0:"
    name="input_:r0:"
    type="text"
  />
</div>
`;
